### PR TITLE
Add pwelch operator

### PR DESCRIPTION
**Pull request by tmartin-gh**
This PR addresses https://github.com/NVIDIA/MatX/issues/279

### CODE REVIEW

File: examples/pwelch.cu [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  auto Pxx  = make_tensor<typename complex::value_type>({nfft});
+
+  for (index_t k = 0; k < signal_size; k++) {
+    // Create the time domain signal as a complex exponential
+    float phase = static_cast<float>(2.0*M_PI*ftone*static_cast<float>(k)/nfft);
+    x(k) = {static_cast<float>(cos(phase)), static_cast<float>(sin(phase))};
+  }
+
+  // Prefetch the data we just created
+  x.PrefetchDevice(0);
```
Comment (cliffburdick) [inline examples/pwelch.cu:79]: In the future we can leave these out of examples since they don't really add anything to it

File: include/matx/operators/pwelch.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+
+#include "matx/core/type_utils.h"
+#include "matx/operators/base_operator.h"
+#include "matx/transforms/pwelch.h"
+
+namespace matx
+{
+  namespace detail {
+    template <typename Op_x>
```
Comment (cliffburdick) [inline include/matx/operators/pwelch.h:43]: Just a minor nit, but all other operators use the convention `OpX`

File: include/matx/operators/pwelch.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    class PWelchOp : public BaseOp<PWelchOp<Op_x>>
+    {
+      private:
+        Op_x x_;
+
+        int nperseg_;
+        int noverlap_;
+        int nfft_;
```
Comment (cliffburdick) [inline include/matx/operators/pwelch.h:51]: Should these be `index_t`?

File: include/matx/operators/pwelch.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        using pwelch_xform_op = bool;
+
+        __MATX_INLINE__ std::string str() const {
+          return "pwelch(" + get_type_str(x_) + ")";
+        }
+
+        __MATX_INLINE__ PWelchOp(const Op_x &x, int nperseg, int noverlap, int nfft) :
+              x_(x), nperseg_(nperseg), noverlap_(noverlap), nfft_(nfft) {
+
+          MATX_ASSERT_STR(x.Rank() == 1, matxInvalidDim, "pwelch:  Only input rank of 1 is supported presently");
```
Comment (cliffburdick) [inline include/matx/operators/pwelch.h:68]: This can use `MATX_STATIC_ASSERT_STR` to do it at compile-time

File: include/matx/transforms/pwelch.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    __MATX_INLINE__ void pwelch_impl(PxxType Pxx, xType x, int nperseg, int noverlap, int nfft, cudaStream_t stream=0)
+    {
+      MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+
+      MATX_ASSERT_STR(Pxx.Rank() == x.Rank(), matxInvalidDim, "pwelch:  Pxx rank must be the same as x rank");
+      MATX_ASSERT_STR(nfft >= nperseg, matxInvalidDim, "pwelch:  nfft must be >= nperseg");
+      MATX_ASSERT_STR((noverlap >= 0) && (noverlap < nperseg), matxInvalidDim, "pwelch:  Must have 0 <= noverlap < nperseg");
+
+      // Create overlapping view
+      auto x_with_overlaps = x.OverlapView({nfft}, {noverlap});
```
Comment (cliffburdick) [inline include/matx/transforms/pwelch.h:66]: This should probably be `overlap` now so you can take in operators and tensors both

File: include/matx/transforms/pwelch.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+      // Create temporary space for fft outputs
+      index_t batches = x_with_overlaps.Shape()[0];
+      auto X_with_overlaps = make_tensor<cuda::std::complex<typename PxxType::scalar_type>>({batches,static_cast<index_t>(nfft)});
```
Comment (cliffburdick) [inline include/matx/transforms/pwelch.h:70]: The last two parameters of `make_tensor` here should be `MATX_ASYNC_DEVICE_MEMORY, stream` so it asynchronously allocates the tensor instead of blocking

File: include/matx/transforms/pwelch.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+      (X_with_overlaps = fft(x_with_overlaps,nfft)).run(stream);
+
+      // Compute magnitude squared in-place
+      (X_with_overlaps = conj(X_with_overlaps) * X_with_overlaps).run(stream);
+      auto mag_sq_X_with_overlaps = X_with_overlaps.RealView();
+
+      // Perform the reduction across 'batches' rows and normalize
+      float norm_factor = 1.f / static_cast<float>(batches);
```
Comment (cliffburdick) [inline include/matx/transforms/pwelch.h:79]: May not be necessary, but `norm_factor` could be cast to `inner_type_t` so it's a double * double or float * float.

File: test/00_transform/PWelch.cu [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+{
+  MATX_ENTER_HANDLER();
+  test.pb->template InitAndRunTVGenerator<TypeParam>(
+      "00_transforms", "pwelch_operators", "pwelch_complex_exponential", {test.params.signal_size, test.params.nperseg, test.params.noverlap, test.params.nfft, test.params.ftone, test.params.sigma});
+
+  tensor_t<TypeParam, 1> x{{test.params.signal_size}};
+  test.pb->NumpyToTensorView(x, "x_in");
+
+  auto Pxx  = make_tensor<typename TypeParam::value_type>({test.params.nfft});
+  (Pxx = pwelch(x, test.params.nperseg, test.params.noverlap, test.params.nfft)).run();
```
Comment (cliffburdick) [inline test/00_transform/PWelch.cu:82]: Can you add:
```
// example-begin pwelch-test-1
 auto Pxx  = make_tensor<typename TypeParam::value_type>({test.params.nfft});
  (Pxx = pwelch(x, test.params.nperseg, test.params.noverlap, test.params.nfft)).run();
  // example-end pwelch-test-1
```

**Comment by cliffburdick** [comment]
Nice! Can you also add the accompanying rst file for documentation? It should have a `pwelch.rst` probably in docs_input/api/signalimage/general unless you think there's a better place. You can copy an existing rst file, then have your example pointed to: `pwelch-test-1`

**Review by cliffburdick — CHANGES_REQUESTED** [review]
Nice! Can you also add an rst in the appropriate directory in docs_input/api pointing to `pwelch-test-1`?

File: include/matx/operators/pwelch.h [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+          Exec(std::make_tuple(tmp_out_), std::forward<Executor>(ex));
+        }
+    };
+  }
+
+  /**
+   *  Estimate the power spectral density of a 1D tensor using Welch's method [1].
+   *
+   * @param x
```
Comment (cliffburdick) [inline include/matx/operators/pwelch.h:115]: Missing `@tparam xType`

File: include/matx/operators/pwelch.h [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+   *   Input time domain tensor x
+   * @param nperseg
+   *   Length of each segment
+   * @param stream
```
Comment (cliffburdick) [inline include/matx/operators/pwelch.h:119]: Stream isn't a parameter here it looks like

File: include/matx/transforms/pwelch.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+/////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+namespace matx
+{
+  /**
+   *  Estimate the power spectral density of a 1D tensor using Welch's method [1].
+   *
+   * @param Pxx
```
Comment (cliffburdick) [inline include/matx/transforms/pwelch.h:40]: Missing `tparam` on `PxxType` and `xType`

File: examples/pwelch.cu [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  cudaStreamCreate(&stream);
+  cudaEvent_t start, stop;
+  cudaEventCreate(&start);
+  cudaEventCreate(&stop);
+
+  // Create buffers
+  auto x  = make_tensor<complex>({signal_size});
+  auto Pxx  = make_tensor<typename complex::value_type>({nfft});
+
+  for (index_t k = 0; k < signal_size; k++) {
```
Comment (luitjens) [inline examples/pwelch.cu:70]: your 2.0 should be 2.0f.
 
 we should also consider an operator implementation.  I don't think we quite have what we need here though.  ideally we would have a sincos operator which takes a real and outputs complex.   @cliffburdick for thoughts.

File: include/matx/operators/pwelch.h [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        index_t noverlap_;
+        index_t nfft_;
+        std::array<index_t, 2> out_dims_;
+        mutable matx::tensor_t<typename OpX::scalar_type, 2> tmp_out_;
+
+      public:
+        using matxop = bool;
+        using scalar_type = typename OpX::scalar_type;
+        using matx_transform_op = bool;
+        using pwelch_xform_op = bool;
```
Comment (luitjens) [inline include/matx/operators/pwelch.h:59]: Why is this necessary?

File: include/matx/operators/pwelch.h [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+        __MATX_INLINE__ std::string str() const {
+          return "pwelch(" + get_type_str(x_) + ")";
+        }
+
+        __MATX_INLINE__ PWelchOp(const OpX &x, index_t nperseg, index_t noverlap, index_t nfft) :
+              x_(x), nperseg_(nperseg), noverlap_(noverlap), nfft_(nfft) {
+
+          MATX_STATIC_ASSERT_STR(x.Rank() == 1, matxInvalidDim, "pwelch:  Only input rank of 1 is supported presently");
+          for (int r = 0; r < x.Rank(); r++) {
```
Comment (luitjens) [inline include/matx/operators/pwelch.h:69]: do we need to store out_dims or can we just query x_ for outdims instead?

File: examples/pwelch.cu (hunk 12 above)
Comment (cliffburdick) [reply examples/pwelch.cu:70]: If you use 2.0f you should also cast M_PI to cast to a float or it's going to do double math anyways. This is just the example setup, so I'm not sure it matters.

Since this is the initialization code, I also don't have a strong opinion on changing this to an operator. If you wanted to, you could make a `sincosf` and `sincospif`  and pass in a `phase` vector to compute it.

File: include/matx/operators/pwelch.h (hunk 13 above)
Comment (cliffburdick) [reply include/matx/operators/pwelch.h:59]: This was a placeholder to use in places where we wanted to detect specific transform operators. It's not used yet, but could come in handy with type traits later for fusion.

File: examples/pwelch.cu (hunk 12 above)
Comment (tmartin-gh) [reply examples/pwelch.cu:70]: I can make the 2.0f and M_PI casting change now.  Do you want to include a complex exponential generator operator in this PR, or leave that for a separate PR?

Comment (tmartin-gh) [reply examples/pwelch.cu]: Pushed a new commit using the existing expj() operator...

**Review by cliffburdick — APPROVED** [review]
